wpilibc/src/test/native/cpp/event/BooleanEventTest.cpp

```diff
@@ -0,0 +1,302 @@
+  EXPECT_EQ(0, counter1);
+  EXPECT_EQ(0, counter2);
+
+  loop.Poll();
+
+  EXPECT_EQ(0, counter1);
+  EXPECT_EQ(0, counter2);
+}
+
+TEST(BooleanEventTest, EdgeDecorators) {
+  EventLoop loop;
+  bool boolean = false;
+  int counter = 0;
+
+  BooleanEvent(&loop, [&] { return boolean; }).Falling().IfHigh([&] {
+    --counter;
+  });
+  BooleanEvent(&loop, [&] { return boolean; }).Rising().IfHigh([&] {
+    ++counter;
+  });
+
+  EXPECT_EQ(0, counter);
+
+  boolean = false;
+  loop.Poll();
+
+  EXPECT_EQ(0, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(1, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(1, counter);
+
+  boolean = false;
+  loop.Poll();
+
+  EXPECT_EQ(0, counter);
+}
+
+/**
+ * Tests that binding actions to the same edge event will result in all actions
+ * executing.
+ */
+TEST(BooleanEventTest, EdgeReuse) {
+  EventLoop loop;
+  bool boolean = false;
+  int counter = 0;
+
+  auto event = BooleanEvent(&loop, [&] { return boolean; }).Rising();
+  event.IfHigh([&] { ++counter; });
+  event.IfHigh([&] { ++counter; });
+
+  EXPECT_EQ(0, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(0, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = false;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(4, counter);
+}
+
+/**
+ * Tests that all actions execute on separate edge events constructed from the
+ * original event.
+ */
+TEST(BooleanEventTest, EdgeReconstruct) {
+  EventLoop loop;
+  bool boolean = false;
+  int counter = 0;
+
+  auto event = BooleanEvent(&loop, [&] { return boolean; });
+  event.Rising().IfHigh([&] { ++counter; });
+  event.Rising().IfHigh([&] { ++counter; });
+
+  EXPECT_EQ(0, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(0, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = false;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(4, counter);
+}
+
+/** Tests that all actions bound to an event will still execute even if the
+ * signal is changed during the loop poll */
+TEST(BooleanEventTest, MidLoopBooleanChange) {
+  EventLoop loop;
+  bool boolean = false;
+  int counter = 0;
+
+  auto event = BooleanEvent(&loop, [&] { return boolean; }).Rising();
+  event.IfHigh([&] {
+    boolean = false;
+    ++counter;
+  });
+  event.IfHigh([&] { ++counter; });
+
+  EXPECT_EQ(0, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(0, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = false;
+  loop.Poll();
+
+  EXPECT_EQ(2, counter);
+
+  boolean = true;
+  loop.Poll();
+
+  EXPECT_EQ(4, counter);
+}
+
+/**
+ * Tests that all actions bound to composed events will still execute even if
+ * the signal is changed during the loop poll.
+ */
+TEST(BooleanEventTest, EventReuse) {
```

Comment:
This test is rather hard to follow. Can you be more descriptive in comments and test name?